### ISSUE TITLE
serious bug in NMS suppression

### ISSUE THREAD

**Issue by Chuong98CC** [ultralytics/ultralytics#22013]
### Search before asking

- [x] I have searched the Ultralytics YOLO [issues](https://github.com/ultralytics/ultralytics/issues) and found no similar bug report.

### Ultralytics YOLO Component

Other

### Bug

```python
import torch
from ultralytics.utils.nms import TorchNMS
# Non-working example
boxes= torch.tensor([[  28.0000, 1322.0000,  346.2500, 1798.0000],
        [ 195,  352,  241,  431],
        [ 194,  352,  241,  431],
        [1160,  545, 1260,  690],
        [1160,  545, 1260,  690],], device='cuda:0')
confidences= torch.tensor([0.6792, 0.6021, 0.5830, 0.5601, 0.5391], device='cuda:0')

keep_indices = TorchNMS.nms(boxes, confidences, iou_threshold=0.2)
print(keep_indices)
# output: tensor([0, 1, 2, 3, 4], device='cuda:0')  --> expect tensor([0, 1, 3], device='cuda:0')

```

### Environment

Ultralytics 8.3.196 🚀 Python-3.10.18 torch-2.8.0+cu128 CUDA:0 (Quadro RTX 8000, 48585MiB)
Setup complete ✅ (20 CPUs, 251.3 GB RAM, 2104.4/3297.7 GB disk)

OS                     Linux-6.5.0-35-generic-x86_64-with-glibc2.35
Environment            Linux
Python                 3.10.18
Install                pip
Path                   /home/chuong/miniconda3/envs/tracktrackcpp/lib/python3.10/site-packages/ultralytics
RAM                    251.34 GB
Disk                   2104.4/3297.7 GB
CPU                    Intel Xeon Silver 4210R 2.40GHz
CPU count              20
GPU                    Quadro RTX 8000, 48585MiB
GPU count              4
CUDA                   12.8

numpy                  ✅ 2.2.6>=1.23.0
matplotlib             ✅ 3.10.6>=3.3.0
opencv-python          ✅ 4.10.0.84>=4.6.0
pillow                 ✅ 11.3.0>=7.1.2
pyyaml                 ✅ 6.0.2>=5.3.1
requests               ✅ 2.32.5>=2.23.0
scipy                  ✅ 1.15.3>=1.4.1
torch                  ✅ 2.8.0>=1.8.0
torch                  ✅ 2.8.0!=2.4.0,>=1.8.0; sys_platform == "win32"
torchvision            ✅ 0.23.0>=0.9.0
psutil                 ✅ 7.0.0
polars                 ✅ 1.33.0
ultralytics-thop       ✅ 2.0.17>=2.0.0

### Minimal Reproducible Example

```python
import torch
from ultralytics.utils.nms import TorchNMS
from torchvision.ops import nms as torch_nms

# Non-working example
boxes= torch.tensor([[  28.0000, 1322.0000,  346.2500, 1798.0000],
        [ 195,  352,  241,  431],
        [ 194,  352,  241,  431],
        [1160,  545, 1260,  690],
        [1160,  545, 1260,  690],], device='cuda:0')
confidences= torch.tensor([0.6792, 0.6021, 0.5830, 0.5601, 0.5391], device='cuda:0')

keep_indices = TorchNMS.nms(boxes, confidences, iou_threshold=0.2)
print(keep_indices)
# output: tensor([0, 1, 2, 3, 4], device='cuda:0')  ---> wrong.

# Working example
keep_indices = torch_nms(boxes, confidences, iou_threshold=0.2)
print(keep_indices)
# output: tensor([0, 1, 3], device='cuda:0')  --> correct

# Working example
boxes= torch.tensor([
        [ 195,  352,  241,  431],
        [ 194,  352,  241,  431],
        [1160,  545, 1260,  690],
        [1160,  545, 1260,  690],], device='cuda:0')
confidences= torch.tensor([ 0.6021, 0.5830, 0.5601, 0.5391], device='cuda:0')

keep_indices = TorchNMS.nms(boxes, confidences, iou_threshold=0.2)
print(keep_indices)
# output: tensor([0, 2], device='cuda:0')  --> correct
```

### Are you willing to submit a PR?

- [ ] Yes I'd like to help by submitting a PR!

**Comment by UltralyticsAssistant** [comment]
👋 Hello @Chuong98CC, thanks for reporting this and for using Ultralytics 🚀. This is an automated response to help you get the fastest possible support while an Ultralytics engineer reviews your issue soon.

We recommend a visit to the [Docs](https://docs.ultralytics.com/) where you can find many [Python](https://docs.ultralytics.com/usage/python/) and [CLI](https://docs.ultralytics.com/usage/cli/) usage examples and where many common questions may already be answered.

Since this is a 🐛 Bug Report:
- Thank you for including a [minimum reproducible example](https://docs.ultralytics.com/help/minimum-reproducible-example/) 🙏
- Please also confirm the behavior on the latest `ultralytics` release and share any additional details that could help us reproduce consistently (e.g., exact inputs, expected vs. actual outputs, and whether it also occurs on CPU)

Join the Ultralytics community where it suits you best. For real-time chat, head to [Discord](https://discord.com/invite/ultralytics) 🎧. Prefer in-depth discussions? Check out [Discourse](https://community.ultralytics.com/). Or dive into threads on our [Subreddit](https://www.reddit.com/r/Ultralytics/) to share knowledge with the community.

## Upgrade

Upgrade to the latest `ultralytics` package including all [requirements](https://github.com/ultralytics/ultralytics/blob/main/pyproject.toml) in a [Python>=3.8](https://www.python.org/) environment with [PyTorch>=1.8](https://pytorch.org/get-started/locally/) to verify your issue is not already resolved in the latest version:

```bash
pip install -U ultralytics
```

## Environments

YOLO may be run in any of the following up-to-date verified environments (with all dependencies including [CUDA](https://developer.nvidia.com/cuda-zone)/[CUDNN](https://developer.nvidia.com/cudnn), [Python](https://www.python.org/) and [PyTorch](https://pytorch.org/) preinstalled):

- **Notebooks** with free GPU: <a href="https://console.paperspace.com/github/ultralytics/ultralytics"><img src="https://assets.paperspace.io/img/gradient-badge.svg" alt="Run on Gradient"/></a> <a href="https://colab.research.google.com/github/ultralytics/ultralytics/blob/main/examples/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/models/ultralytics/yolo11"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>
- **Google Cloud** Deep Learning VM. See [GCP Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/google_cloud_quickstart_tutorial/)
- **Amazon** Deep Learning AMI. See [AWS Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/aws_quickstart_tutorial/)
- **Docker Image**. See [Docker Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/docker_image_quickstart_tutorial/) <a href="https://hub.docker.com/r/ultralytics/ultralytics"><img src="https://img.shields.io/docker/pulls/ultralytics/ultralytics?logo=docker" alt="Docker Pulls"></a>

If possible, please try reproducing your example in one of the environments above and share the notebook or exact steps you used. This helps us isolate environment-specific issues faster ⚙️

## Status

<a href="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule"><img src="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml/badge.svg" alt="Ultralytics CI"></a>

If this badge is green, all [Ultralytics CI](https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule) tests are currently passing. CI tests verify correct operation of all YOLO [Modes](https://docs.ultralytics.com/modes/) and [Tasks](https://docs.ultralytics.com/tasks/) on macOS, Windows, and Ubuntu every 24 hours and on every commit.

Additional details that will help us investigate:
- Exact `ultralytics` version after upgrade and confirmation of the behavior on CPU vs. CUDA
- Any warnings or logs printed when running your MRE
- Whether the issue appears when running in a clean environment (e.g., Colab) using the same snippet

Thanks again for the detailed report—an Ultralytics engineer will follow up shortly 🧪

**Comment by glenn-jocher** [comment]
Thanks for the clear MRE—reproduced; the issue is in the early-termination path of `TorchNMS.nms` where we mistakenly keep all remaining boxes when the top-scoring box has zero overlap, skipping NMS among the rest; see the function in the [TorchNMS.nms reference](https://docs.ultralytics.com/reference/utils/nms/#nms).

Hotfix you can apply locally in `ultralytics/utils/nms.py`:
```python
# Replace this block in TorchNMS.nms(...)
# if inter.sum() == 0:
#     remaining_count = rest.numel()
#     keep[keep_idx : keep_idx + remaining_count] = rest
#     keep_idx += remaining_count
#     break

# with this:
if inter.sum() == 0:
    order = rest
    continue
```

Workaround until the fix lands: keep using `torchvision.ops.nms` (as in your MRE) or import `torchvision` before running inference so `non_max_suppression` selects the torchvision NMS path automatically, as shown in the [non_max_suppression docs](https://docs.ultralytics.com/reference/utils/nms/#non_max_suppression). Please confirm this patch returns `tensor([0, 1, 3], device='cuda:0')` on your example so we can close the loop here.

**Comment by Chuong98CC** [comment]
yes, it works, thanks for the hotfix

**Comment by glenn-jocher** [comment]
Thanks for confirming—we’ll submit a PR with this fix and add a regression test; for future reports, the MRE guide is helpful: [Creating a Minimal Reproducible Example](https://docs.ultralytics.com/help/minimum-reproducible-example/).

**Comment by UltralyticsAssistant** [comment]
A potential fix for this issue has been merged in the pull request Fix TorchNMS.nms() early exit logic by @Y-T-G with contributions from @Laughing-q. The PR improves NMS correctness and stability, which should address cases where overlapping boxes were incorrectly kept.

Key changes that may resolve the issue:
- Replaced `scores.sort(...)[1]` with `scores.argsort(...)` to get index order directly, simplifying and slightly improving efficiency.
- Corrected the early-exit logic: when the top-scoring box has no overlap with the remaining boxes, the loop now continues processing instead of leaving the rest unchecked. This reduces false positives and improves detection reliability across models like YOLO11.
- Minor cleanup of masking and comments for clarity and maintainability.

Testing options:
- Install the latest main branch to verify the fix immediately:
  - `pip install -U git+https://github.com/ultralytics/ultralytics.git@main`
- Alternatively, wait for the next official PyPI release and then:
  - `pip install -U ultralytics`
- After updating, rerun your original reproduction. For a quick sanity check:
  ```python
  from ultralytics import YOLO

  model = YOLO("yolo11n.pt")  # or your model
  results = model("path/to/your/test/image_or_video")
  # Review detections to confirm improved NMS behavior
  ```

Feedback request:
- Please let us know if the merged changes resolve the issue in your environment. If problems persist, sharing a minimal repro (model, data sample, and command) will help with further investigation.

Thanks for reporting this and helping improve the project. 🙏
